backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/matchers/IrFunctionMatcher.kt

```diff
@@ -0,0 +1,112 @@
+package org.jetbrains.kotlin.backend.konan.lower.matchers
+
+import org.jetbrains.kotlin.backend.konan.irasdescriptors.fqNameSafe
+import org.jetbrains.kotlin.ir.declarations.IrFunction
+import org.jetbrains.kotlin.ir.declarations.IrValueParameter
+import org.jetbrains.kotlin.name.FqName
+
+
+internal inline fun createFunctionMatcher(restrictions: IrFunctionMatcher.() -> Unit): IrFunctionMatcher =
+        IrFunctionMatcher().apply(restrictions)
+
+data class ParameterRestriction(
+        val index: Int,
+        val restriction: (IrValueParameter) -> Boolean
+)
+
+internal class IrFunctionMatcher(var functionKind: Kind = Kind.ANY) {
+
+    enum class Kind {
+        ANY,
+        NO_RECEIVER,
+        EXTENSION,
+        METHOD
+    }
+
+    private val receiverRestrictions = mutableListOf<(IrValueParameter) -> Boolean>()
+
+    private val parameterRestrictions = mutableListOf<ParameterRestriction>()
+
+    private val nameRestrictions = mutableListOf<(FqName) -> Boolean>()
+
+    private val paramCountRestrictions = mutableListOf<(Int) -> Boolean>()
+
+    fun fqNameRestriction(restriction: (FqName) -> Boolean) {
+        nameRestrictions += restriction
+    }
+
+    fun parameterRestriction(index: Int, restriction: (IrValueParameter) -> Boolean) {
+        parameterRestrictions += ParameterRestriction(index, restriction)
+    }
+
+    fun receiverRestriction(restriction: (IrValueParameter) -> Boolean) {
+        receiverRestrictions += restriction
+    }
+
+    fun parametersSizeRestriction(restriction: (Int) -> Boolean) {
```

Comment:
There is no such thing as "size of parameters", but there is "count of parameters".